src/coreclr/vm/threads.cpp

```diff
@@ -7475,6 +6974,119 @@ PTR_GCFrame Thread::GetGCFrame()
+        {
+            NOTHROW;
+            GC_TRIGGERS;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        {
+            GCX_PREEMP();
+
+            ForeignThreadsToCleanUpEvent->Wait(INFINITE, FALSE);
+        }
+
+        if (Thread::m_DetachCount > 0)
+        {
+            Thread::CleanupDetachedThreads();
+        }
+
+        ThreadStore::s_pThreadStore->TriggerGCForDeadThreadsIfNecessary();
+    }
+
+    DWORD WINAPI ForeignThreadCleanupThreadStart(LPVOID lpParameter)
+    {
+        bool cleanupThreadOk = ((Thread*)lpParameter)->HasStarted();
+
+        _ASSERTE(cleanupThreadOk);
+
+        // finalizer should always park in default domain
+
+        if (cleanupThreadOk)
+        {
+            INSTALL_UNHANDLED_MANAGED_EXCEPTION_TRAP;
+            {
+                while (!CleanupThreadShouldStop)
+                {
+                    ManagedThreadBase::KickOff(ForeignThreadCleanupWorker, NULL);
+                }
+                _ASSERTE(g_fEEShutDown);
+                CleanupThreadShutdownEvent->Set();
+            }
+            UNINSTALL_UNHANDLED_MANAGED_EXCEPTION_TRAP;
+        }
+        return 0;
+    }
+}
+
+void
+ThreadCleanupThread::EnsureCleanupThreadExists()
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    } CONTRACTL_END;
+
+    ForeignThreadsToCleanUpEvent = new CLREvent();
+    ForeignThreadsToCleanUpEvent->CreateAutoEvent(FALSE);
+    CleanupThreadShutdownEvent = new CLREvent();
+    CleanupThreadShutdownEvent ->CreateAutoEvent(FALSE);
+
+    Thread* pCleanupThread = SetupUnstartedThread();
+
+    if (pCleanupThread->CreateNewThread(0, &ForeignThreadCleanupThreadStart, pCleanupThread, W(".NET External Thread Cleanup Thread")))
+    {
+        DWORD dwRet = pCleanupThread->StartThread();
+
+        // When running under a user mode native debugger there is a race
+        // between the moment we've created the thread (in CreateNewThread) and
+        // the moment we resume it (in StartThread); the debugger may receive
+        // the "ct" (create thread) notification, and it will attempt to
+        // suspend/resume all threads in the process.  Now imagine the debugger
+        // resumes this thread first, and only later does it try to resume the
+        // newly created thread (the finalizer thread).  In these conditions our
+        // call to ResumeThread may come before the debugger's call to ResumeThread
+        // actually causing dwRet to equal 2.
+        // We cannot use IsDebuggerPresent() in the condition below because the
+        // debugger may have been detached between the time it got the notification
+        // and the moment we execute the test below.
+        _ASSERTE(dwRet == 1 || dwRet == 2);
+    }
+}
+void
+ThreadCleanupThread::SetHasThreadsToCleanUp()
+{
+    _ASSERT(ForeignThreadsToCleanUpEvent != nullptr);
+    ForeignThreadsToCleanUpEvent->Set();
+}
+
+void
+ThreadCleanupThread::ShutdownCleanupThread()
```

Comment:
Why do we need this? I would expect that the cleanup thread can stay around and be killed by the OS.